### PR TITLE
Rewrite the "Composition Concepts" documentation

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/2470

I recently had cause to write a bunch of new XRs and Compositions as part of https://github.com/crossplane/crossplane/issues/2568. I was frustrated by how often I needed to go read the code to remember how things worked, so I decided to try to make the Composition documentation a little more detailed. I've hopefully also made it a little less impenetrable by breaking it up into concepts and a reference (previously it was a bit of both).

As pertains to Composition, this PR:

* Simplifies the Composition diagrams (and makes them prettier).
* Attempts to have the Concepts page start with the basics and build on them.
* Breaks all the deeper details out into a new 'reference' page.
* Add tips, tricks and troubleshooting to the reference page.

I've left the 'getting started' guide out of scope for this change, since I know we've had a lot of discussion around how to improve that. This PR does however contain a few other small tweaks (all in separate commits). I believe the only one that might be in any way controversial is that I'm trying to undo the "Composite resource kinds start with `Composite`" convention I established back in the day and replace it with "Composite resource kinds start with `X`". So:

* Before a `PostgreSQLInstance` claim corresponded to a `CompositePostgreSQLInstance` XR (by convention).
* Now a `PostgreSQLInstance` claim corresponds to an `XPostgreSQLInstance` XR (still by convention).

My rational is that `X` is way less to type, and seems to visually make it easier to match up the two related types. It also sounds cooler.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

I've done one proof reading pass. You can find a build of this branch at https://negz.github.io/crossplane.github.io/docs/. See in particular:

* [The "Composite Resources" page](https://negz.github.io/crossplane.github.io/docs/v1.4/concepts/composition.html) under "Concepts".
* [The "Composition" page](https://negz.github.io/crossplane.github.io/docs/v1.4/reference/composition.html) under "Reference".

### CODE REVIEW

**Comment by negz** [comment]
CC @luebken, @vfarcic, and @stevendborrelli who are not (yet) Crossplane maintainers but are likely to be interested in this change.

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:62]: I would add the abbrevations XRD, XR and MR to the diagram.

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:64]: Another note I would move. ;-)

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:74]: An XRD is a ~little like~ CRD.

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:112]: The content is great but it doesn't need to be a note.

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:203]: No real suggestion here just a comment that I had to read it twice to understand this section.

File: docs/concepts/composition.md
Comment (luebken) [inline docs/concepts/composition.md:220]: Why are there two XPostgreSQLInstances but only one XNetwork?

Comment (luebken) [reply docs/concepts/composition.md:220]: Oh. It belongs to the scenario described below. I wonder how we could make this more clear. E.g. move the diagram to the bottom?

**Review by luebken — COMMENTED** [review]
Great additions and very helpful. 

Possibly for a second polishing pass but I think we can improve on the diagram style.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md]: I was on the fence about this but held off because historically we rarely abbreviate Managed Resource to MR. I don't see why we wouldn't though - I updated this paragraph to introduce the abbreviation but have not refactored to use it elsewhere. Most of `managed-resources-md` is just two documents I merged together, so I would prefer to avoid too much diff on the actual content in this PR.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md]: Done.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md:22]: Most of `managed-resources.md` is just two documents I merged together. I would prefer to make any non-trivial updates to the content of `managed-resources.md` in a future PR.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md:43]: Most of `managed-resources.md` is just two documents I merged together. I would prefer to make any non-trivial updates to the content of `managed-resources.md` in a future PR.

That said, my personal preference is to keep the details on how to install providers under the section where that already exists - I think it would quickly become repetitive and difficult to maintain if we tried to supply all the prior context needed to get to a state inline.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md:76]: Most of `managed-resources.md` is just two documents I merged together. I would prefer to make any non-trivial updates to the content of `managed-resources.md` in a future PR.

FWIW I'd like to remove the synced condition per https://github.com/crossplane/crossplane-runtime/issues/198.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md:280]: This seems quite subjective. Personally I prefer the note to 'interrupt' me closer to where it's relevant, rather than being a footnote at the end of the document that I may never encounter if I'm opening a heading to get some specific targeted context about how something works rather than reading the entire document end-to-end.

File: docs/concepts/managed-resources.md
Comment (negz) [reply docs/concepts/managed-resources.md:292]: As per previous comments, I didn't touch this content in this PR. I'd prefer to leave it out of scope for a future PR.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:62]: I'd prefer not to. I think using the full names does enough to explain what each concept is, and feel adding the abbreviations would clutter it visually.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:74]: That's not correct.

File: docs/concepts/composition.md
Comment (stevendborrelli) [inline docs/concepts/composition.md:176]: should the provider-kubernetes and provider-helm links resolve to the GitHub repositories?

File: docs/concepts/composition.md
Comment (stevendborrelli) [reply docs/concepts/composition.md:62]: I showed this document [how-it-works] to some new Crossplane users, and they really liked it! However, as we discussed it we felt making it clear which components are defined by the Platform Team, and what is generated dynamically by Crossplane could help. 

Could we emphasize in the diagram that the CompositeResource is a new k8s object generated within crossplane from the XR + Composition?

**Review by stevendborrelli — APPROVED** [review]

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:64]: I think this one makes sense as a quote note. It's at the end of the heading so I don't feel it interrupts.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:203]: I was on the fence about including this at all - I had copied it from the terminology guide when I was trying to frame up the claims section. I've removed part of it that I didn't think was very useful, and merged the "public and private" analogy part into the opening paragraph.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:112]: De-noted.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:18]: I removed this one.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:243]: That's good to hear. It's historically been quite an undocumented/underutilised edge case.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:176]: They do for me. These are "footnote" style markdown links, which I feel makes the raw markdown more readable. See the `provider-kubernetes` link in the list of links at the bottom.

**Comment by negz** [comment]
> I think we can improve on the diagrams a bit. Style and annotations. Possibly for a second polishing pass.

Do you have any actionable feedback apart from that which you left inline? You're welcome to comment directly on the sources at https://drive.google.com/drive/u/2/folders/1ld-XNglPQBT4ue9IBVuDs4HQRizr-5AY if it helps.

> I haven't reviewed the reference yet but my previous comments around too many quote notes also apply there. e.g. if you look at this section: https://negz.github.io/crossplane.github.io/docs/v1.4/reference/composition.html#compositions There is no visual guidance and the user can't scan the text as there are too many different styles.

I can't help but feel this is a highly personal/subjective thing. I personally like quote notes and read them as little "side note" style commentary. I prefer this to having to (e.g.) find some other part of the document such as footnotes that are likely away from where I am reading. I personally don't open up a documentation site and read it end-to-end but rather skim for the parts that are relevant to the question I have, so having clarifications nearby helps. That said, I have tried to take a critical eye and "demoted" _some_ of the quote notes to hopefully make the document flow better for those who dislike them.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:61]: > Could we emphasize in the diagram that the CompositeResource is a new k8s object generated within crossplane from the XR + Composition?

I think there might be a terminology mixup - this doesn't seem like an accurate representation of what happens.

In an ideal world we want to communicate:

1. A **type of** Composite Resource exists because it was defined by a Composite Resource Definition.
2. Ditto for a type of Claim.
3. An **instance of** a Claim exists when an app team member creates it.
4. An **instance of** a Composite Resource exists either:
     a. Because it was created automatically by Crossplane in response to a Claim.
     b. Because a platform team member created it.
5. An instance of a Composite Resource uses a Composition to determine what Managed Resources to compose.
6. Instances of Managed Resources exist because a Composite Resource created them (per its Composition)

My hunch is that this is way too much to fit in a diagram, and that we should have a slightly 'low resolution' diagram that makes more sense when you read the surrounding context (i.e. documentation).

That said, I'm open to suggestions for light weight changes we could make here. Perhaps we could:

* Add a "_Creates_ Composite Resource" label to the Claim -> Composite Resource relationship.
* Add a "_Created_ by Composite Resource" subtitle to the Managed Resource cluster.
* Change the Composite Resource Definition subtitle to "_Defines_ Composite Resource and Claim Types".
* Change the Composition subtitle to "_Template_ for Managed Resources"

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md:215]: It does illustrate that scenario, but I hope it can be a broader demonstration of claims. I experimented with a few places to position the diagram but feel best about where it is. Generally I like the flow of diagram first, then text that explains the diagram. I've added some more labels to this one that hopefully makes what is going on a little clearer. Let me know whether they help.

**Comment by negz** [comment]
I've addressed most of the feedback here and pushed an updated build of the documentation to https://negz.github.io/crossplane.github.io/docs/.

File: docs/concepts/composition.md
Comment (luebken) [reply docs/concepts/composition.md]: Regarding the position of the diagram: LGTM
The diagram transports a lot of information now. I think it would be helpful to remove certain aspects. e.g. the responsibilities. And leave that to the explanation in the text.

**Comment by luebken** [comment]
> I can't help but feel this is a highly personal/subjective thing. I personally like quote notes and read them as little "side note" style commentary. I prefer this to having to (e.g.) find some other part of the document such as footnotes that are likely away from where I am reading. I personally don't open up a documentation site and read it end-to-end but rather skim for the parts that are relevant to the question I have, so having clarifications nearby helps. That said, I have tried to take a critical eye and "demoted" some of the quote notes to hopefully make the document flow better for those who dislike them.

OK. But readability/scanability of product documentation is not a subjective thing buts it's hard to quantify. At this point, it was only meant as a hint. I can't justify this. This would need to be done by some UX expert. But outside of this PR.

File: docs/concepts/composition.md
Comment (negz) [reply docs/concepts/composition.md]: Makes sense - done.

**Comment by negz** [comment]
I'd like to move forward with merging this PR if folks are 90+% on board with the changes, but I need an approve from a maintainer. @muvaf @jbw976 I think you would be great folks to review this as subject matter experts on Composition.

**Review by muvaf — APPROVED** [review]
LGTM! I love the new `Composite Resources` under `Concepts`, we can finally point to an end to end composition guide that's not as comprehensive as the reference page. Thank you @negz !

**Pull request merged**
